File: gcj2010R2A.py
#!/usr/bin/env python

def solve((k,K,d)):
    def dist(i): return abs(k-(i+1))
    def elegant(i,j):
        def inside(i,j): return 0<=i and i<K and dist(i)<=j and j<K-dist(i)
        for x,y in [(x,y) for x in range(K) for y in range(dist(x),K-dist(x))]:
            for x1,y1 in [(x,j*2-y),(i*2-x,y)]:
                if inside(x1,y1) and d[x][y]!=d[x1][y1]:
                    return False
        return True
    def diff(i,j): return (k+sum(map(dist,[i,j])))**2-k**2
    return min([diff(i,j) for i in range(K) for j in range(K) if elegant(i,j)])

if __name__ == '__main__':
    import sys
    from multiprocessing import Pool
    def getin():
        k = int(sys.stdin.readline())
        K = k*2-1
        return (k,K,[sys.stdin.readline().rstrip() for _ in range(K)])
    xs = [getin() for _ in range(int(sys.stdin.readline()))]
    for i,x in enumerate(Pool().imap(solve,xs)):
        print("Case #"+str(i+1)+": "+str(x))

File: gcj2010R2A.cpp
#include <iostream>
#include <vector>
#include <string>
#include <limits>

static auto solve = []()
{
    int k; std::cin >> k;
    const auto K = 2*k-1;
    { std::string t; std::getline(std::cin,t); }
    std::vector<std::string> d(K);
    for(auto& v: d) std::getline(std::cin,v);
    static auto inside = [&](const int x, const int y)
    {
        const auto yb=std::abs(k-(x+1)), ye=K-yb;
        return (0<=x && x<K) && (yb<=y && y<ye);
    };
    static auto elegant = [&](const int i, const int j)
    {
        for(int x=0; x<K; ++x)
        {
            for(int y=0; y<K; ++y)
            {
                if(inside(x,y))
                {
                    const auto x1=i*2-x, y1=j*2-y;
                    if( (inside(x,y1) && d[x][y]!=d[x][y1]) ||
                        (inside(x1,y) && d[x][y]!=d[x1][y]) )
                        return false;
                }
            }
        }
        return true;
    };
    int n = std::numeric_limits<int>::max();
    for(int i=0; i<K; ++i)
    {
        for(int j=0; j<K; ++j)
        {
            if(elegant(i,j))
            {
                const int d = std::abs(k-(i+1))+std::abs(k-(j+1));
                n=std::min(n,(k+d)*(k+d)-k*k);
            }
        }
    }
    return n;
};

int main()
{
    int T; std::cin >> T;
    for(int i=1; i<=T; ++i)
        std::cout << "Case #" << i << ": " << solve() << '\n';
}
